### PR TITLE
wireless/bcm43xxx: correct auth status if PSK is invaild

### PR DESCRIPTION
**Pull request by anchao**
## Summary

wireless/bcm43xxx: discard auth event if netdev down
wireless/bcm43xxx: subscribe only to the events we need
wireless/bcm43xxx: skip WEP privacy
wireless/bcm43xxx: skip bad channel bss
wireless/bcm43xxx: correct auth status if PSK is invaild
wireless/bcm43xxx: only report PSK ssid by scan filter

## Impact

N/A

## Testing

bcm43013

### CODE REVIEW

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 1]
```diff
@@ -114,6 +123,48 @@ enum
...
+  uint8_t tag;                  /* TAG */
+  uint8_t length;               /* TAG length */
+  uint8_t oui[3];               /* IE OUI */
+  uint8_t oui_type;             /* OUI type */
+  begin_packed_struct struct
+    {
+      uint8_t low;
+      uint8_t high;
+    }
+  end_packed_struct version;  /* IE version */
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:163]:
```suggestion
  end_packed_struct version;    /* IE version */
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 2]
```diff
@@ -501,14 +549,6 @@ int bcmf_wl_active(FAR struct bcmf_dev_s *priv, bool active)
...
 int bcmf_driver_initialize(FAR struct bcmf_dev_s *priv)
 {
-  int i;
-
-  /* FIXME Configure event mask to enable all asynchronous events */
-
-  for (i = 0; i < BCMF_EVENT_COUNT; i++)
-    {
-      bcmf_event_register(priv, bcmf_wl_default_event_handler, i);
-    }
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:552]:
```suggestion
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 3]
```diff
@@ -695,6 +762,128 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
+                  /* WPA_OUI */
+
+                  if (memcmp(&ie->oui[0], "\x00\x50\xf2\x01", 4))
+                    {
+                      break;
+                    }
+
+                  vaild_bss = false;
+
+                  rsn = (wpa_rsn_t *)&ie_buffer[ie_offset +
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:850]:
```suggestion
                  rsn = (FAR wpa_rsn_t *)&ie_buffer[ie_offset +
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 4]
```diff
@@ -695,6 +762,128 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
+                              goto vaild_bss;
+                            }
+                        }
+                    }
+                  break;
+                }
+
+              case IEEE80211_ELEMID_VENDOR:
+                {
+                  FAR wpa_ie_fixed_t *ie = (wpa_ie_fixed_t *)
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:836]:
```suggestion
                  FAR wpa_ie_fixed_t *ie = (FAR wpa_ie_fixed_t *)
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 5]
```diff
@@ -578,11 +605,15 @@ void bcmf_wl_auth_event_handler(FAR struct bcmf_dev_s *priv,
                                 struct bcmf_event_s *event,
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:605]:
```suggestion
                                FAR struct bcmf_event_s *event,
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 6]
```diff
@@ -124,9 +175,6 @@ static int bcmf_driver_download_clm(FAR struct bcmf_dev_s *priv);
 
 /* FIXME only for debug purpose */
 
-static void bcmf_wl_default_event_handler(FAR struct bcmf_dev_s *priv,
-                            struct bcmf_event_s *event, unsigned int len);
-
 static void bcmf_wl_radio_event_handler(FAR struct bcmf_dev_s *priv,
                             struct bcmf_event_s *event, unsigned int len);
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:179]:
```suggestion
                            FAR struct bcmf_event_s *event, unsigned int len);
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 7]
```diff
@@ -1720,10 +1897,9 @@ int bcmf_wl_get_rssi(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
 
 int bcmf_wl_set_encode_ext(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1898]:
```suggestion
int bcmf_wl_set_encode_ext(FAR struct bcmf_dev_s *priv, FAR struct iwreq *iwr)
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 8]
```diff
@@ -1720,10 +1897,9 @@ int bcmf_wl_get_rssi(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
...
 {
-  int interface;
   struct iw_encode_ext *ext;
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1900]:
```suggestion
  FAR struct iw_encode_ext *ext;
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 9]
```diff
@@ -1766,11 +1943,12 @@ int bcmf_wl_set_encode_ext(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
 
 int bcmf_wl_set_ssid(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1944]:
```suggestion
int bcmf_wl_set_ssid(FAR struct bcmf_dev_s *priv, FAR struct iwreq *iwr)
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 10]
```diff
@@ -1779,39 +1957,64 @@ int bcmf_wl_set_ssid(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
       return -EINVAL;
     }
 
-  if (iwr->u.essid.flags)
+  out_len = sizeof(scbval);
+  memset(&scbval, 0x0, out_len);
+  ret = bcmf_cdc_ioctl(priv, interface, true,
+                       WLC_DISASSOC, (uint8_t *)&scbval, &out_len);
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1963]:
```suggestion
                       WLC_DISASSOC, (FAR uint8_t *)&scbval, &out_len);
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 11]
```diff
@@ -1779,39 +1957,64 @@ int bcmf_wl_set_ssid(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
...
+    }
 
-      out_len = sizeof(ssid);
+  if (priv->auth_pending)
+    {
+      out_len = sizeof(priv->auth_pmk);
       ret = bcmf_cdc_ioctl(priv, interface, true,
-                         WLC_SET_SSID, (uint8_t *)&ssid, &out_len);
+                           WLC_SET_WSEC_PMK,
+                           (uint8_t *)&priv->auth_pmk, &out_len);
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1974]:
```suggestion
                           (FAR uint8_t *)&priv->auth_pmk, &out_len);
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 12]
```diff
@@ -1779,39 +1957,64 @@ int bcmf_wl_set_ssid(FAR struct bcmf_dev_s *priv, struct iwreq *iwr)
...
+
+  ssid.ssid_len = iwr->u.essid.length;
+  memcpy(ssid.SSID, iwr->u.essid.pointer, iwr->u.essid.length);
+
+  /* Configure AP SSID and trig authentication request */
+
+  out_len = sizeof(ssid);
+  ret = bcmf_cdc_ioctl(priv, interface, true,
+                       WLC_SET_SSID,
+                       (uint8_t *)&ssid, &out_len);
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:2004]:
```suggestion
                       (FAR uint8_t *)&ssid, &out_len);
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.h [hunk 13]
```diff
@@ -92,8 +92,10 @@ struct bcmf_dev_s
   FAR wl_bss_info_t *scan_result;      /* Temp buffer that holds results */
   unsigned int scan_result_entries;    /* Current entries of temp buffer */
 
-  sem_t auth_signal; /* Authentication notification signal */
-  int   auth_status; /* Authentication status */
+  sem_t *auth_signal;   /* Authentication notification signal */
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.h:95]:
```suggestion
  FAR sem_t *auth_signal;   /* Authentication notification signal */
```

Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.h:95]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 12 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:2004]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 11 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1974]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 10 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1963]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 9 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1944]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 8 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:1900]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 6 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:179]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 5 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:605]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 1 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:163]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 2 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:552]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 3 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:850]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 4 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:836]: Done

**Comment by anchao** [comment]
@pkarashchenko  all addressed, thank you!

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**